### ISSUE TITLE
Equation for binary_backward is not the same in REBAR

### ISSUE THREAD

**Issue by backpropper** [Bonnevie/rebar#3]
The function binary_backward defined [here](https://github.com/Bonnevie/rebar/blob/0ff4715e8f17d970fdcff57fc195b984d551e799/relaxflow/reparam.py#L164) does not use the same equation as what is written in the REBAR paper (Appendix C).

**Comment by Bonnevie** [comment]
The reparameterization might be slightly different, but it should be correct. `p` is not the *p*robability, but the odds *p*arameter `np.log(prob/(1-prob))`.
Also, it doesn't really matter what reparameterization you use as long as it is consistent.
edit: In other words, it has not been a design goal to make the reparameterization match the REBAR paper's choice exactly. The reparameterization I use is in terms of an unconstrained variable, and is numerically stable.

**Comment by backpropper** [comment]
I see. I guess `p` is defined as `log alpha` which is the output of the inference network where `theta = sigmoid(log alpha)` is the probability and `alpha = theta / (1- theta)`.

What I meant was that your reparameterization is the same for b=1 case but different for the b=0 case.

**Comment by Bonnevie** [comment]
`uprime = tf.nn.sigmoid(-p)` gives the bernoulli probability, which splits the [0,1] interval into the `b=0` and `b=1` case. The `ub` line defines a conditional uniform distribution on either the interval [u',1] or [0,u'], depending on `b`.  The last equation with `zb` corresponds to the original REBAR reparameterization, with the conditional `ub` draw plugged in.

So it should be correct? Could you be more specific about your concerns? How is it different for b=0?

**Comment by backpropper** [comment]
I am saying that if you plug in the values I wrote in the equations given in the Appendix C of the REBAR paper you will reach the same form as yours for b=1 case but not for b=0 case.

**Comment by Bonnevie** [comment]
Double-checked it now. The reason you get the difference is that the dependency on the uniform noise `v` is flipped. So `v` sweeps out the same set of reparameterized states, but in opposite order. So you can substitute `v` for `1-v` and you should get the same result.

The appendix was somewhat different back when I wrote the code, just to give some background for the differences. That aside, I sort of dislike their way of reparameterizing, it's very obtuse.

**Comment by backpropper** [comment]
You mean substitute just for the b=0 case. Yes that works. You might wanna change the code?
I totally agree that their form of reparameterization is obscure and even they use a very different approach in their code. Your approach seems to be more intuitive including the coupling method.

**Comment by Bonnevie** [comment]
Do you think it's necessary to switch? I feel like the code as currently implemented is more intuitive, as you say, and it's equally correct. Or is it causing you problems?

**Comment by backpropper** [comment]
No the code is working fine. Its just that since you use the name REBAR, it would match the equations given in the paper.

**Comment by backpropper** [comment]
The coupling function is not the same as [here](https://github.com/tensorflow/models/blob/26aded58ab0f8c08698b41db3958d9ca54707fe2/research/rebar/rebar.py#L438) (the original implementation). Can you share your thoughts on using a different approach? (Ignore the last line in the function that should not have been there and George confirmed it)

Actually, your interpretation makes more sense.

**Comment by Bonnevie** [comment]
Hmm, mechanically it is the same. My implementation is directly inspired by theirs. Mostly just tried to clean it up a bit since it seemed unnecessarily obtuse. Also, mine depends on `b` directly which I hope is more convenient, while their version checks `u>u_prime`. I also need my implementation to fit into the general extendable interface, so that coupling can be generalized to other distributions without having to reimplement the estimator.

**Comment by backpropper** [comment]
Actually the equations are not the same even when you convert yours to check the condition `u>u_prime`. They convert `v` into `u` space which enables them to use the same reparameterization function as `u`. Their equations are:
`v_1= (u-up) / (1-up) * (1-up) + up`
and 
`v_0 = u / up * up`
How do you interpret that?

**Comment by Bonnevie** [comment]
hm, actually, that seems a bit odd. The terms cancel in their equations, leaving `v_1=u` and `v_0=u`, which will not lead to correct coupling unless they have made some changes upstream. If you look at my [binary_backward](https://github.com/Bonnevie/rebar/blob/0ff4715e8f17d970fdcff57fc195b984d551e799/relaxflow/reparam.py#L164) and then plug in the binary [coupling](https://github.com/Bonnevie/rebar/blob/0ff4715e8f17d970fdcff57fc195b984d551e799/relaxflow/reparam.py#L147) noise `v`, you can see that `ub` (the conditional noise in `binary_backward`) reduces to `u` as desired in my version.

**Comment by backpropper** [comment]
They have a stop gradient in the first fractional part before the multiplication.
So in your case zb=z in the forward pass but they are different in the backward pass? If this is true then their equatuon is correct given that they also use the same reparameterization function for both u and v.

**Comment by Bonnevie** [comment]
Hmm, I think you've spent more time looking at their implementation recently than I have so please correct me if wrong, but the implementation seems a bit weird - it looks like they are using the same function `_random_sample_soft` to sample from both the conditional `p(z)` and the unconditional `p(z|b)`? This could of course just be a refactoring, but it means that they have to move the differentiable bits (e.g. the map from uniform noise `v` to conditional noise `ub`) of the conditional sampler to the definition of `v`, which is likely why they mess around with the `stop_gradients` in `_u_to_v`. My code is distributed differently, with `coupling` defining the map from `u` to `v` (which is wrapped in `stop_gradient` by the `reparam` class), and `binary_backward` defining the map from `v` to `ub` and from `ub` to `z`. So you need to take my `coupling`, wrap it in `stop_gradient`, and plug it into `binary_backward` to compare.

I feel like they are kinda going against their own definitions by having `v` in the code (in `_u_to_v` for instance) actually be the full conditional coupling `ub`. Basically, their `v` is equal to my `ub`.

**Comment by backpropper** [comment]
Thanks for the detailed explanation. Yes I agree their implementation seems a bit convoluted.
